evm/vm_state.py

```diff
@@ -0,0 +1,325 @@
+from contextlib import contextmanager
+import copy
+import logging
+
+from cytoolz import (
+    merge,
+)
+from eth_utils import (
+    encode_hex,
+)
+
+from evm.constants import (
+    MAX_PREV_HEADER_DEPTH,
+)
+from evm.db.tracked import (
+    AccessLogs,
+)
+from evm.exceptions import (
+    BlockNotFound,
+)
+from evm.utils.state import (
+    make_trie_root_and_nodes,
+)
+
+
+class BaseVMState(object):
+    #
+    # Set from __init__
+    #
+    _chaindb = None
+    block_header = None
+    prev_headers = None
+
+    computation_class = None
+    access_logs = None
+    receipts = None
+
+    def __init__(self, chaindb, block_header, prev_headers, receipts=[]):
+        self._chaindb = chaindb
+        self.block_header = block_header
+        self.prev_headers = prev_headers
+
+        self.access_logs = AccessLogs()
+        self.receipts = receipts
+
+    #
+    # Logging
+    #
+    @property
+    def logger(self):
+        return logging.getLogger('evm.vm_state.{0}'.format(self.__class__.__name__))
+
+    #
+    # Block Object Properties (in opcodes)
+    #
+    @property
+    def blockhash(self):
+        return self.block_header.hash
+
+    @property
+    def coinbase(self):
+        return self.block_header.coinbase
+
+    @property
+    def timestamp(self):
+        return self.block_header.timestamp
+
+    @property
+    def block_number(self):
+        return self.block_header.block_number
+
+    @property
+    def difficulty(self):
+        return self.block_header.difficulty
+
+    @property
+    def gas_limit(self):
+        return self.block_header.gas_limit
+
+    #
+    # chaindb
+    #
+    # @property
+    # def chaindb(self):
+    #     return self._chaindb
+
+    def set_chaindb(self, db):
+        self._chaindb = db
+
+    #
+    # state_db
+    #
+    @contextmanager
+    def state_db(self, read_only=False):
+        state = self._chaindb.get_state_db(self.block_header.state_root, read_only)
+        yield state
+
+        if read_only:
+            # This acts as a secondary check that no mutation took place for
+            # read_only databases.
+            assert state.root_hash == self.block_header.state_root
+        elif self.block_header.state_root != state.root_hash:
+            self.block_header.state_root = state.root_hash
+
+        self.access_logs.reads.update(state.db.access_logs.reads)
+        self.access_logs.writes.update(state.db.access_logs.writes)
+
+        # remove the reference to the underlying `db` object to ensure that no
+        # further modifications can occur using the `State` object after
+        # leaving the context.
+        state.db = None
+        state._trie = None
```

Comment:
Note to self:

There was a `decomission()` method added to the state database object to remove the need to access this private variable.  We should backport that improvement from the sharding branch.